Updates manual integration tests with documentation

joaomcteixeira opened the pull request:
You are about to submit a new Pull Request. Before continuing make sure you read the [contributing guidelines](CONTRIBUTING.md) and you comply with the following criteria:

- [x] Your PR is about writing documentation for already existing code :fire:
- [x] Your PR is about writing tests for already existing code :godmode:

---

Because more of us are now participating in the development of haddock3, I push my scripts to run the integration tests manually. This is how I am doing, but I think they can become part of the repository so that everyone has the guidelines and the tools to use them. Please read the documentation to see if it is clear.

Added more `-test.cfg` examples for antibody-antigen cases.

Added `ncores=40` to all tests. Anyway, haddock internally will decide which is the best `ncores` value.

Added a `run-test.job` to send tests to slurm.

Even when we automate theses tests, having the capacity to run them locally is always good.

mgiulini reviewed (commented):
Very nice @joaomcteixeira. I just added two minor comments.

joaomcteixeira commented on docs/integration_tests.md:
````diff
@@ -0,0 +1,95 @@
+## 2. Install HADDOCK3 main branch in a separate folder and python environment
+
+```bash
+git clone https://github.com/haddocking/haddock3 haddock3main
````
Totally right!

joaomcteixeira commented:
Done it @mgiulini :+1: thanks for your feedback

rvhonorato commented on docs/integration_tests.md:
```diff
@@ -0,0 +1,103 @@
+# Running integration tests manually
+
+The integration tests ensure HADDOCK3 workflows work properly by
+performing short docking simulations on benchmark systems. With these
```
Not everything is docking and these are not benchmark systems

joaomcteixeira replied:
for docking :+1:
Okay, I can reword the "benchmark". But a *benchmark* is anything used as a reference. These are benchmarks because they are used as test cases. But I will reword it because HADDOCK uses "benchmark" for a very specific purpose. I think you were referring to that. :+1:

rvhonorato commented on examples/diff_capri.sh:
```diff
@@ -0,0 +1,73 @@
+# HADDOCK main relative path
+HM="../../haddock3main/examples"
+
+diff_capri () {
```
Why is this being done with shell and not with python? Can't we parse these and compare if the values are within a tolerable range? How can the developer tell if the numbers "match"?

joaomcteixeira replied:
The quick answer is that I just wanted to share my scripts with the main dev team. Since we are not yet on an official release. I can make it on Python, but for testing, is that needed?

Regarding comparing the values within a tolerable range. Do we want a tolerable range for integration tests? Remember tests are run on the same machines and the numbers should be equal.

The developer knows the numbers match because `diff` gives no output. Otherwise `diff` shows the different lines.

rvhonorato replied:
> I can make it on Python, but for testing, is that needed?

"The project guidelines" say haddock3 is python exclusive (: 

> Remember tests are run on the same machines and the numbers should be equal.

So maybe I did not understand what this part is doing, is it not comparing values obtained by the tests with an expected one?

joaomcteixeira replied:
Okay, better make it in python. It might be useful for the future.

> So maybe I did not understand what this part is doing, is it not comparing values obtained by the tests with an expected one?

No, is just using the bash `diff` command. If files are equal, no output. If numbers are different, files are different, `diff` outputs the difference. Just that.

rvhonorato commented on examples/run-tests.job:
```diff
@@ -0,0 +1,16 @@
+cd $HOME/haddock3/examples
+
+# execute
+python run_tests.py -b
```
I'd change this to be a template instead, this is very particular to our cluster - other's will not have a `short` queue. Best to avoid people just blindly trying to run things.

joaomcteixeira replied:
Yeah. Okay. I wrote in the docs the developer should edit it. But I can make it a template. But what parts should I transform the template? Should I remove the `queue` line?

rvhonorato replied:
Is this is just for internal usage I'd actually remove it. People running HPC jobs are advanced users and can write simple scripts

joaomcteixeira replied:
Yes. It is as stated in the documentation. But, yes. Maybe is better to remove that mention and the `run-test.job`.

rvhonorato commented on docs/integration_tests.md:
```diff
@@ -0,0 +1,103 @@
+
+## 5. Comparing both runs
+
+Once both runs completed, that is, all `*-test.cfg` files were executed,
+you can compare the different results for the CAPRI scoring modules.
```
Sorry, now I understand: you make a branch, make modifications, compare yours with the main and check the diff, ok.

But remember that due to the non-deterministic nature of haddock, changes in the CNS level might produce models with slight variations and these will be propagated to the capri metrics (ex 0.147 instead of 0.152). This is what I meant about values being within a certain range.

joaomcteixeira replied:
Yes. I actually suggest keeping two separate haddock3 installations: one for development and the other for the `main`, just for this testing purposes. That's how I am doing.

Regarding the numbers, when running on the same machine, there should be not these different numbers, I think. So far, numbers have been always the same in all the PR tests I have made in the last months.

Obviously, if the changes affect the CAPRI scores on purpose, we cannot use these as validity metrics.

I will have a look further at your comments and edit accordingly.

rvhonorato replied:
> Regarding the numbers, when running on the same machine, there should be not these different numbers, I think. 

You are wrong. I'm speaking from experience, but fine - we can cross that bridge when the times comes to it.

joaomcteixeira replied:
Even for 3 decimal numbers? Okay. I can make it to test `close` numbers, that's not a problem. I was planning to make in Python as you suggested.

rvhonorato replied:
As I said already, its not numerical fluctuation, its a result of a **slightly different model**, but nevermind that, let's not worry with future problems hehe

joaomcteixeira replied:
Is not a future problem, is understanding haddock's behavior. So far, I see deterministic results: that is, running the examples several times always gives the exact same numbers. I am taking note of your comments :+1:

rvhonorato commented:
Also, since you are at it, is it not better if we move all the test-related scripts to the `tests/` folder instead? I think its a bit odd that we have examples and test together

joaomcteixeira commented:
> Also, since you are at it, is it not better if we move all the test-related scripts to the tests/ folder instead? I think its a bit odd that we have examples and tests together.

To what I explained in the ISGC meeting, and I am writing it now in the form of text, I would say it is okay to have tests in the `examples/`. The `tests/` folder are for unnittests or API tests. And in this case the `examples/*/*-test.cfg` are integration tests. I would keep them separated.

The `-test.cfg` in `examples/` also serves to demonstrate to users before running the actual `-full.cfg`. We had discussed this before, as of today, I still agree with this rationale. However, your comment is important. :+1:

amjjbonvin commented:
In this context benchmark should be replaced by test systems 

The goal is not to get the best results but to ensure the workflow works properly

joaomcteixeira commented:
> In this context benchmark should be replaced by test systems. The goal is not to get the best results but to ensure the workflow works properly

Understood, done it :+1:

joaomcteixeira commented:
Okay, ready for review:

- Made the comparison script in Python
- Update the docs. Are they clear?
- Added three more `*-test.cfg` cases for `antibody-antigen`
- added tests
- removed the `.job` file. Agree with @rvhonorato; advanced users should know what to do.

amjjbonvin commented on docs/integration_tests.md:
```diff
@@ -0,0 +1,113 @@
+stop if any of the tests break. For help, see `python
+run_tests.py -h`.
+
+## 4. Running the tests on your branch
```
Is this the usual way to do it?  I am creating branches in my main installation and switching between branches.

joaomcteixeira replied:
That's also possible. If you have two separate installations, you can run the tests simultaneously: main and developing. I thought about it, but I preferred to document the two installations because I think it is easier than asking to copy the `examples/` folder to a new one and be switching between branches. But both ways work.

joaomcteixeira commented on examples/docking-antibody-antigen/docking-antibody-antigen-CDR-accessible-clt-test.cfg:
```diff
@@ -0,0 +1,95 @@
+# ====================================================================
+# Protein-protein docking example with NMR-derived ambiguous interaction restraints
```
@mgiulini implementation was not merged yet -> #397
Whoever gets merged first, the other needs to update :wink:

joaomcteixeira commented on docs/integration_tests.md:
```diff
@@ -0,0 +1,113 @@
+stop if any of the tests break. For help, see `python
+run_tests.py -h`.
+
+## 4. Running the tests on your branch
```
I use slurm batch submission to run the `*-test.cfg`, and I have the `job` files for that in my account. Initially, I had an explanation for using Slurm, but removed it because I agreed with @rvhonorato comments (https://github.com/haddocking/haddock3/pull/396#discussion_r849123487). But I can add a comment for that now. Should I?

joaomcteixeira replied:
Added short message in the docs

joaomcteixeira commented:
> "- if I just want to test if proteinA interacts with proteinB, should I use docking-protein-protein-mdref-full.cfg or docking-protein-protein-mdref-test.cfg?"

@rvhonorato 

Added option to `compare_runs.py` to compare two specific run directories.

mgiulini commented on docs/integration_tests.md:
```diff
@@ -0,0 +1,130 @@
+**How to correctly set up and run the integration tests?**
+
+The easiest way is to have two installations of haddock3:
+one fixed at the `main` branch and other with your developments.
```
```suggestion
one fixed at the `main` branch and another with your developments.
```

mgiulini commented on docs/integration_tests.md:
```diff
@@ -0,0 +1,130 @@
+Navigate to your development repository and haddock3 environment - you
+are likely on a different terminal window.
+
+Run the tests the same way:
+
+Navigate to the `examples/` folder and execute the tests:
```
I would lump these two sentences in a single one

mgiulini commented on examples/compare_runs.py:
```diff
@@ -0,0 +1,347 @@
+printf = partial(print, flush=True)  # noqa: T002
+
+
+examples = (
+    ("docking-antibody-antigen"    , "docking-antibody-antigen-ranairCDR-test.cfg"),  # noqa: E203, E501
+    ("docking-antibody-antigen"    , "docking-antibody-antigen-ranairCDR-clt-test.cfg"),  # noqa: E203, E501
+    ("docking-antibody-antigen"    , "docking-antibody-antigen-CDR-accessible-test.cfg"),  # noqa: E203, E501
+    ("docking-antibody-antigen"    , "docking-antibody-antigen-CDR-accessible-clt-test.cfg"),  # noqa: E203, E501
+    ("docking-protein-DNA"         , "docking-protein-DNA-test.cfg"),  # noqa: E203, E501
+    ("docking-protein-DNA"         , "docking-protein-DNA-mdref-test.cfg"),  # noqa: E203, E501
+    ("docking-protein-homotrimer"  , "docking-protein-homotrimer-test.cfg"),  # noqa: E203, E501
+    ("docking-protein-ligand-shape", "docking-protein-ligand-shape-test.cfg"),  # noqa: E203, E501
+    ("docking-protein-ligand"      , "docking-protein-ligand-test.cfg"),  # noqa: E203, E501
+    ("docking-protein-peptide"     , "docking-protein-peptide-test.cfg"),  # noqa: E203, E501
+    ("docking-protein-peptide"     , "docking-protein-peptide-mdref-test.cfg"),  # noqa: E203, E501
+    ("docking-protein-protein"     , "docking-protein-protein-test.cfg"),  # noqa: E203, E501
+    ("docking-protein-protein"     , "docking-protein-protein-cltsel-test.cfg"),  # noqa: E203, E501
+    ("docking-protein-protein"     , "docking-protein-protein-mdref-test.cfg"),  # noqa: E203, E501
+    ("refine-complex"              , "refine-complex-test.cfg"),  # noqa: E203, E501
+    ("scoring"                     , "emscoring-test.cfg"),  # noqa: E203, E501
+    ("scoring"                     , "mdscoring-test.cfg"),  # noqa: E203, E501
+    )
```
I am a bit skeptical about hard-coding all these examples here, as we have them hard-coded in another script (`run_tests.py`). Wouldn't be better to create a list and import it in both cases? so that when we add a an example we only need to modify the file with the list

joaomcteixeira replied:
Yes, you are right. These were hard coded at first when we were prototyping them. But now it makes little sense. I will arrange that in another pull request.

joaomcteixeira commented:
Yes, updated with all your comments. I think it is fine now. `compare_runs.py` can be used to compare `examples` folders as a whole or individual run directories.
